tfjs-core/src/io/weights_loader.ts

```diff
@@ -245,3 +235,180 @@ export function weightsLoaderFactory(
     return weightsTensorMap;
   };
 }
+
+type BufferRange = {
+  start: number,
+  end: number,
+  buffer: ArrayBuffer,
+};
+
+export class CompositeArrayBuffer {
+  private ranges: BufferRange[] = [];
+  private previousRangeIndex = 0;
+  private bufferUniformSize?: number;
+  public readonly byteLength: number;
+
+  constructor(buffers: ArrayBuffer | ArrayBuffer[] | TypedArray
+      | TypedArray[]) {
+    // Normalize the `buffers` input to be `ArrayBuffer[]`.
+    if (!(buffers instanceof Array)) {
+      buffers = [buffers];
+    }
+    buffers = buffers.map((bufferOrTypedArray) => {
+      if (util.isTypedArray(bufferOrTypedArray)) {
+        return bufferOrTypedArray.buffer;
+      }
+      return bufferOrTypedArray;
+    });
+
+    // Skip setting up ranges if there are no buffers.
+    if (buffers.length === 0) {
+      return;
+    }
+
+    this.bufferUniformSize = buffers[0].byteLength;
+    let start = 0;
+
+    for (let i = 0; i < buffers.length; i++) {
+      const buffer = buffers[i];
+      // Check that all buffers except the last one have the same length.
+      if (i !== buffers.length - 1 &&
+        buffer.byteLength !== this.bufferUniformSize) {
+        // Unset the buffer uniform size, since the buffer sizes are not
+        // uniform.
+        this.bufferUniformSize = undefined;
+      }
+
+      // Create the ranges, including their start and end points.
+      const end = start + buffer.byteLength;
+      this.ranges.push({buffer, start, end,});
+      start = end;
+    }
+
+    // Set the byteLenghth
+    if (this.ranges.length === 0) {
+      this.byteLength = 0;
+    }
+    this.byteLength = this.ranges[this.ranges.length - 1].end;
+  }
+
+  slice(start = 0, end = this.byteLength): ArrayBuffer {
+    // NaN is treated as zero for slicing. This matches ArrayBuffer's behavior.
```

Comment:
convert start and end to Number with `Number(...)` before checking NaN?

Reply:
Update:
I assume you add these nan checks because you think there may be calls from JS now or future which ignores the typescript type check. In these way I'd suggest to do `start = Number(start)` since `isNaN('123')` returns `true` and `ArrayBuffer.prototype.slice` accepts the numbers in strings.

Reply:
I added these NaN checks because some of the tests were failing (they intentionally gave it no datatype, so I think the tests themselves are correct). I'd like this to match `ArrayBuffer.slice` as closely as possible, so I [implemented your comment.](https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/io/composite_array_buffer.ts#L89-L90)